scipy/integrate/_quadrature.py

```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    spaced, then the result is exact only if the function is a polynomial of 
+    order 2 or less.
+
+    References
+    ----------
+    .. [1] Wikipedia page: https://en.wikipedia.org/wiki/Simpson's_rule
+    .. [2] Cartwright, Kenneth V. Simpson's Rule Cumulative Integration with
+            MS Excel and Irregularly-spaced Data. Journal of Mathematical
+            Sciences and Mathematics Education. 12 (2): 1-9
+
+    Examples
+    --------
+    >>> from scipy import integrate
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> x = np.linspace(-2, 2, num=20)
+    >>> y = x**2
+    >>> y_int = integrate.cumulative_simpson(y, x=x, initial=0)
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(x, y_int, 'ro', x, x**3/3 - (x[0])**3/3, 'b-')
+    >>> ax.grid()
+    >>> plt.show()
+
+    """
+    y = np.asarray(y)
+
+    # validate y and axis
+    if axis < -1 or axis >= y.ndim:
+        raise ValueError("If given, `axis` must exist in the shape of `y`.")
+
+    if y.shape[axis] < 3:
+        raise ValueError(
+            "At least 3 points are required along the axis of integration "
+            "to use the composite Simpson's method."
+        )
+
+    if x is not None:
+        x = np.asarray(x)
+        if x.ndim == 1 and y.shape[axis] == x.shape[0]:
+            x_shape = [1] * y.ndim
+            x_shape[axis] = -1
+            x = x.reshape(tuple(x_shape))
+        elif x.shape != y.shape:
+            raise ValueError(
+                "If given, shape of `x` must be the same as `y` or 1-D with "
+                "the same length as `y` along `axis`."
+            )
+        res = _cumulative_simpson_unequal_intervals(y, x, axis=axis)
+
+    else:
+        dx = np.asarray(dx)
+        final_dx_shape = tupleset(y.shape, axis, y.shape[axis] - 1)
+        alt_input_dx_shape = tupleset(y.shape, axis, 1)
+
+        if dx.ndim == 0:
+            dx = np.ones(final_dx_shape) * dx
+        elif dx.shape == alt_input_dx_shape:
+            dx = np.repeat(dx, y.shape[axis] - 1, axis=axis)
+        else:
+            raise ValueError(
+                "`dx` must either be numeric or have the same shape as `y` "
+                "but with only 1 point along `axis`."
+            )
+        res = _cumulative_simpson_equal_intervals(y, dx, axis=axis)
+
+    if initial is not None:
+        initial = np.asarray(initial)
+        alt_initial_input_shape = tupleset(y.shape, axis, 1)
+        if initial.ndim == 0:
+            initial = np.ones(y.shape) * initial
+        elif initial.shape == alt_initial_input_shape:
+            initial = np.repeat(initial, y.shape[axis], axis=axis)
+        else:
+            raise ValueError(
+                "`initial` must either be numeric or have the same shape as "
+                "`y` but with only 1 point along `axis`."
+            )
+
+        slice1 = tupleset((slice(None),) * y.ndim, axis, slice(1, None))
+        initial[slice1] = initial[slice1] + res
+        res = initial
```

Comment:
I think we can simplify this a lot with the help of `np.broadcast_arrays`. To make sure I understand what you're after, is this exactly true?
> `dx` must either be numeric or have the same shape as `y` but with only 1 point along `axis`.

To be clear, suppose `x` and `y` are valid inputs. Then you would not allow the user to use:
```
dx = np.diff(x)
```
as `dx` instead of passing `x`, right?

Reply:
Your understanding is correct regarding dx. It is consistent with cumulative_trapezoid.

I agree with you though, the np.broadcast* functions should be used to simplify this. Will get back to you after this updates.

Reply:
Well... it would have really simplified things if the predecessors followed normal broadcasting rules, but it doesn't. So I'd rather just make this one consistent with the other three. I'd suggest factoring out the array processing code from `cumulative_trapezoid` into a function and using it in both functions.